In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
from google.colab import files

uploaded = files.upload()

Saving airbnb.xlsx to airbnb.xlsx


In [4]:
import io

file_name = next(iter(uploaded))

df = pd.read_excel(io.BytesIO(uploaded[file_name]))

print("Dataset loaded successfully!")
print("File:", file_name)

Dataset loaded successfully!
File: airbnb.xlsx


In [5]:
print("Dataset Shape:", df.shape)
display(df.head())

Dataset Shape: (48895, 16)


,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaT,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0


In [6]:
print("Dataset Shape:", df.shape)

print("\nColumn Names:")
print(df.columns.tolist())

print("\nData Types:")
print(df.dtypes)

Dataset Shape: (48895, 16)

Column Names:
['id', 'name', 'host_id', 'host_name', 'neighbourhood_group', 'neighbourhood', 'latitude', 'longitude', 'room_type', 'price', 'minimum_nights', 'number_of_reviews', 'last_review', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365']

Data Types:
id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                      int64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calc

In [7]:
print("Missing Values in Each Column:")
print(df.isnull().sum())

print("\nTotal Missing Values:")
print(df.isnull().sum().sum())

print("\nDuplicate Rows:")
print(df.duplicated().sum())

Missing Values in Each Column:
id                                    0
name                                 26
host_id                               0
host_name                            22
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

Total Missing Values:
20152

Duplicate Rows:
0


In [8]:
print("Statistical Summary:")
display(df.describe().T)

Statistical Summary:


,count,mean,min,25%,50%,75%,max,std
id,48895.0,19017143.23618,2539.0,9471945.0,19677284.0,29152178.5,36487245.0,10983108.38561
host_id,48895.0,67620010.64661,2438.0,7822033.0,30793816.0,107434423.0,274321313.0,78610967.032667
latitude,48895.0,40.728949,40.49979,40.6901,40.72307,40.763115,40.91306,0.05453
longitude,48895.0,-73.95217,-74.24442,-73.98307,-73.95568,-73.936275,-73.71299,0.046157
price,48895.0,152.720687,0.0,69.0,106.0,175.0,10000.0,240.15417
minimum_nights,48895.0,7.029962,1.0,1.0,3.0,5.0,1250.0,20.51055
number_of_reviews,48895.0,23.274466,0.0,1.0,5.0,24.0,629.0,44.550582
last_review,38843,2018-10-04 01:47:23.910099456,2011-03-28 00:00:00,2018-07-08 00:00:00,2019-05-19 00:00:00,2019-06-23 00:00:00,2019-07-08 00:00:00,NaN
reviews_per_month,38843.0,1.373221,0.01,0.19,0.72,2.02,58.5,1.680442
calculated_host_listings_count,48895.0,7.143982,1.0,1.0,1.0,2.0,327.0,32.952519


In [9]:
categorical_columns = [
    'neighbourhood_group',
    'neighbourhood',
    'room_type'
]

for column in categorical_columns:
    print(f"\n{column}")
    print("Unique values:", df[column].nunique())
    print(df[column].unique())


neighbourhood_group
Unique values: 5
['Brooklyn' 'Manhattan' 'Queens' 'Staten Island' 'Bronx']

neighbourhood
Unique values: 221
['Kensington' 'Midtown' 'Harlem' 'Clinton Hill' 'East Harlem'
 'Murray Hill' 'Bedford-Stuyvesant' "Hell's Kitchen" 'Upper West Side'
 'Chinatown' 'South Slope' 'West Village' 'Williamsburg' 'Fort Greene'
 'Chelsea' 'Crown Heights' 'Park Slope' 'Windsor Terrace' 'Inwood'
 'East Village' 'Greenpoint' 'Bushwick' 'Flatbush' 'Lower East Side'
 'Prospect-Lefferts Gardens' 'Long Island City' 'Kips Bay' 'SoHo'
 'Upper East Side' 'Prospect Heights' 'Washington Heights' 'Woodside'
 'Brooklyn Heights' 'Carroll Gardens' 'Gowanus' 'Flatlands' 'Cobble Hill'
 'Flushing' 'Boerum Hill' 'Sunnyside' 'DUMBO' 'St. George' 'Highbridge'
 'Financial District' 'Ridgewood' 'Morningside Heights' 'Jamaica'
 'Middle Village' 'NoHo' 'Ditmars Steinway' 'Flatiron District'
 'Roosevelt Island' 'Greenwich Village' 'Little Italy' 'East Flatbush'
 'Tompkinsville' 'Astoria' 'Clason Point' 'East

In [10]:
numerical_columns = [
    'latitude',
    'longitude',
    'price',
    'minimum_nights',
    'number_of_reviews',
    'reviews_per_month',
    'calculated_host_listings_count',
    'availability_365'
]

for column in numerical_columns:
    print(f"\n--- {column} ---")
    print("Minimum:", df[column].min())
    print("Maximum:", df[column].max())
    print("Negative values:", (df[column] < 0).sum())
    print("Zero values:", (df[column] == 0).sum())


--- latitude ---
Minimum: 40.49979
Maximum: 40.91306
Negative values: 0
Zero values: 0

--- longitude ---
Minimum: -74.24442
Maximum: -73.71299
Negative values: 48895
Zero values: 0

--- price ---
Minimum: 0
Maximum: 10000
Negative values: 0
Zero values: 11

--- minimum_nights ---
Minimum: 1
Maximum: 1250
Negative values: 0
Zero values: 0

--- number_of_reviews ---
Minimum: 0
Maximum: 629
Negative values: 0
Zero values: 10052

--- reviews_per_month ---
Minimum: 0.01
Maximum: 58.5
Negative values: 0
Zero values: 0

--- calculated_host_listings_count ---
Minimum: 1
Maximum: 327
Negative values: 0
Zero values: 0

--- availability_365 ---
Minimum: 0
Maximum: 365
Negative values: 0
Zero values: 17533


In [11]:
zero_price = df[df['price'] == 0]

print("Number of listings with zero price:", len(zero_price))

display(zero_price)

Number of listings with zero price: 11


,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
23161,18750597,"Huge Brooklyn Brownstone Living, Close to it all.",8993084,Kimberly,Brooklyn,Bedford-Stuyvesant,40.69023,-73.95428,Private room,0,4,1,2018-01-06,0.05,4,28
25433,20333471,★Hostel Style Room | Ideal Traveling Buddies★,131697576,Anisha,Bronx,East Morrisania,40.83296,-73.88668,Private room,0,2,55,2019-06-24,2.56,4,127
25634,20523843,"MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)",15787004,Martial Loft,Brooklyn,Bushwick,40.69467,-73.92433,Private room,0,2,16,2019-05-18,0.71,5,0
25753,20608117,"Sunny, Quiet Room in Greenpoint",1641537,Lauren,Brooklyn,Greenpoint,40.72462,-73.94072,Private room,0,2,12,2017-10-27,0.53,2,0
25778,20624541,Modern apartment in the heart of Williamsburg,10132166,Aymeric,Brooklyn,Williamsburg,40.70838,-73.94645,Entire home/apt,0,5,3,2018-01-02,0.15,1,73
25794,20639628,Spacious comfortable master bedroom with nice ...,86327101,Adeyemi,Brooklyn,Bedford-Stuyvesant,40.68173,-73.91342,Private room,0,1,93,2019-06-15,4.28,6,176
25795,20639792,Contemporary bedroom in brownstone with nice view,86327101,Adeyemi,Brooklyn,Bedford-Stuyvesant,40.68279,-73.91170,Private room,0,1,95,2019-06-21,4.37,6,232
25796,20639914,Cozy yet spacious private brownstone bedroom,86327101,Adeyemi,Brooklyn,Bedford-Stuyvesant,40.68258,-73.91284,Private room,0,1,95,2019-06-23,4.35,6,222
26259,20933849,the best you can find,13709292,Qiuchi,Manhattan,Murray Hill,40.75091,-73.97597,Entire home/apt,0,3,0,NaT,NaN,1,0
26841,21291569,Coliving in Brooklyn! Modern design / Shared room,101970559,Sergii,Brooklyn,Bushwick,40.69211,-73.90670,Shared room,0,30,2,2019-06-22,0.11,6,333


In [12]:
high_min_nights = df[df['minimum_nights'] > 365]

print("Listings with minimum_nights > 365:", len(high_min_nights))

display(high_min_nights[['id', 'name', 'host_name', 'minimum_nights', 'price', 'room_type']])

Listings with minimum_nights > 365: 14


,id,name,host_name,minimum_nights,price,room_type
1305,568684,800sqft apartment with huge terrace,Alessandra,370,115,Entire home/apt
2854,1615764,NaN,Peter,1000,400,Entire home/apt
5767,4204302,Prime W. Village location 1 bdrm,Genevieve,1250,180,Entire home/apt
7355,5431845,Beautiful Fully Furnished 1 bed/bth,Aliya,500,134,Entire home/apt
8014,6169897,Wonderful Large 1 bedroom,John,500,75,Entire home/apt
10829,8341919,Brand New Luxury Apt Lease Takeover,Shining,480,199,Entire home/apt
11193,8668115,Zen Room in Crown Heights Brooklyn,Laura,500,50,Private room
13404,10053943,Historic Designer 2 Bed. Apartment,Glenn H.,999,99,Entire home/apt
14285,11096888,Peaceful apartment close to F/G,Amanda,500,45,Private room
15946,12888849,"Bright Luxury Studio, Central Loc.",Sarah,366,250,Entire home/apt


In [13]:
Q1 = df['minimum_nights'].quantile(0.25)
Q3 = df['minimum_nights'].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

outliers_min_nights = df[
    (df['minimum_nights'] < lower_bound) |
    (df['minimum_nights'] > upper_bound)
]

print("\nNumber of IQR outliers:", len(outliers_min_nights))

Q1: 1.0
Q3: 5.0
IQR: 4.0
Lower Bound: -5.0
Upper Bound: 11.0

Number of IQR outliers: 6607


In [14]:
display(
    outliers_min_nights[
        ['id', 'name', 'minimum_nights', 'price', 'room_type']
    ].sort_values('minimum_nights', ascending=False).head(20)
)

,id,name,minimum_nights,price,room_type
5767,4204302,Prime W. Village location 1 bdrm,1250,180,Entire home/apt
2854,1615764,NaN,1000,400,Entire home/apt
26341,20990053,Beautiful place in Brooklyn! #2,999,79,Private room
13404,10053943,Historic Designer 2 Bed. Apartment,999,99,Entire home/apt
38664,30378211,Shared Studio (females only),999,110,Shared room
47620,35834935,Williamsburg Apartment,500,140,Entire home/apt
8014,6169897,Wonderful Large 1 bedroom,500,75,Entire home/apt
7355,5431845,Beautiful Fully Furnished 1 bed/bth,500,134,Entire home/apt
14285,11096888,Peaceful apartment close to F/G,500,45,Private room
11193,8668115,Zen Room in Crown Heights Brooklyn,500,50,Private room


In [15]:
print("Minimum Nights Percentiles:")
print(df['minimum_nights'].quantile([0.50, 0.75, 0.90, 0.95, 0.99, 0.995, 1.00]))

Minimum Nights Percentiles:
0.500       3.0
0.750       5.0
0.900      28.0
0.950      30.0
0.990      45.0
0.995      90.0
1.000    1250.0
Name: minimum_nights, dtype: float64


In [16]:
zero_price = df[df['price'] == 0]

print("Number of zero-price listings:", len(zero_price))

display(
    zero_price[
        ['id', 'name', 'host_name', 'price', 'minimum_nights', 'room_type']
    ]
)

Number of zero-price listings: 11


,id,name,host_name,price,minimum_nights,room_type
23161,18750597,"Huge Brooklyn Brownstone Living, Close to it all.",Kimberly,0,4,Private room
25433,20333471,★Hostel Style Room | Ideal Traveling Buddies★,Anisha,0,2,Private room
25634,20523843,"MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)",Martial Loft,0,2,Private room
25753,20608117,"Sunny, Quiet Room in Greenpoint",Lauren,0,2,Private room
25778,20624541,Modern apartment in the heart of Williamsburg,Aymeric,0,5,Entire home/apt
25794,20639628,Spacious comfortable master bedroom with nice ...,Adeyemi,0,1,Private room
25795,20639792,Contemporary bedroom in brownstone with nice view,Adeyemi,0,1,Private room
25796,20639914,Cozy yet spacious private brownstone bedroom,Adeyemi,0,1,Private room
26259,20933849,the best you can find,Qiuchi,0,3,Entire home/apt
26841,21291569,Coliving in Brooklyn! Modern design / Shared room,Sergii,0,30,Shared room


In [17]:
rows_before_cleaning = len(df)

df = df[df['price'] > 0].copy()

rows_after_price_cleaning = len(df)

print("Rows before removing zero prices:", rows_before_cleaning)
print("Rows after removing zero prices:", rows_after_price_cleaning)
print("Rows removed:", rows_before_cleaning - rows_after_price_cleaning)

Rows before removing zero prices: 48895
Rows after removing zero prices: 48884
Rows removed: 11


In [18]:
missing_values = df.isnull().sum()

print("Missing Values After Price Cleaning:")
print(missing_values[missing_values > 0])

Missing Values After Price Cleaning:
name                    26
host_name               22
last_review          10051
reviews_per_month    10051
dtype: int64


In [19]:
print("Missing name:", df['name'].isnull().sum())
print("Missing host_name:", df['host_name'].isnull().sum())

display(
    df[df['name'].isnull() | df['host_name'].isnull()]
    [['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
      'neighbourhood', 'room_type', 'price']]
)

Missing name: 26
Missing host_name: 22


,id,name,host_id,host_name,neighbourhood_group,neighbourhood,room_type,price
360,100184,Bienvenue,526653,NaN,Queens,Queens Village,Private room,50
2700,1449546,Cozy Studio in Flatbush,7779204,NaN,Brooklyn,Flatbush,Entire home/apt,100
2854,1615764,NaN,6676776,Peter,Manhattan,Battery Park City,Entire home/apt,400
3703,2232600,NaN,11395220,Anna,Manhattan,East Village,Entire home/apt,200
5745,4183989,SPRING in the City!! Zen-Style Tranquil Bedroom,919218,NaN,Manhattan,Harlem,Private room,86
5775,4209595,NaN,20700823,Jesse,Manhattan,Greenwich Village,Entire home/apt,225
5975,4370230,NaN,22686810,Michaël,Manhattan,Nolita,Entire home/apt,215
6075,4446862,Charming Room in Prospect Heights!,23077718,NaN,Brooklyn,Crown Heights,Private room,50
6269,4581788,NaN,21600904,Lucie,Brooklyn,Williamsburg,Private room,150
6567,4756856,NaN,1832442,Carolina,Brooklyn,Bushwick,Private room,70


In [21]:
# Check whether missing reviews_per_month corresponds to listings with no reviews

missing_review_data = df[df['reviews_per_month'].isnull()]

print("Missing reviews_per_month:", len(missing_review_data))
print("Number of reviews for these listings:")
print(missing_review_data['number_of_reviews'].value_counts())

Missing reviews_per_month: 10051
Number of reviews for these listings:
number_of_reviews
0    10051
Name: count, dtype: int64


In [22]:
# Check the relationship between the two review-related missing columns

print("Missing last_review:", df['last_review'].isnull().sum())
print("Missing reviews_per_month:", df['reviews_per_month'].isnull().sum())

print("\nRows where reviews_per_month is missing but number_of_reviews > 0:")
display(
    df[
        df['reviews_per_month'].isnull() &
        (df['number_of_reviews'] > 0)
    ][['id', 'number_of_reviews', 'last_review', 'reviews_per_month']]
)

Missing last_review: 10051
Missing reviews_per_month: 10051

Rows where reviews_per_month is missing but number_of_reviews > 0:


,id,number_of_reviews,last_review,reviews_per_month


In [23]:
# Handle missing values

df['reviews_per_month'] = df['reviews_per_month'].fillna(0)

df['name'] = df['name'].fillna('Unknown')

df['host_name'] = df['host_name'].fillna('Unknown')

print("Missing values after handling:")
display(df.isnull().sum()[df.isnull().sum() > 0])

Missing values after handling:


,0
last_review,10051


In [24]:
# Convert last_review to datetime

df['last_review'] = pd.to_datetime(df['last_review'], errors='coerce')

print("Data type of last_review:")
print(df['last_review'].dtype)

print("\nMissing values in last_review:")
print(df['last_review'].isnull().sum())

Data type of last_review:
datetime64[ns]

Missing values in last_review:
10051


In [25]:
# Check unique values in important categorical columns

print("Neighbourhood Groups:")
print(df['neighbourhood_group'].unique())

print("\nRoom Types:")
print(df['room_type'].unique())

print("\nNumber of unique neighbourhoods:", df['neighbourhood'].nunique())

Neighbourhood Groups:
['Brooklyn' 'Manhattan' 'Queens' 'Staten Island' 'Bronx']

Room Types:
['Private room' 'Entire home/apt' 'Shared room']

Number of unique neighbourhoods: 221


In [26]:
# Check for leading/trailing whitespace

for col in ['name', 'host_name', 'neighbourhood_group', 'neighbourhood', 'room_type']:
    whitespace_count = (
        df[col].dropna()
        .astype(str)
        .apply(lambda x: x != x.strip())
        .sum()
    )
    print(f"{col}: {whitespace_count} values with extra whitespace")

name: 239 values with extra whitespace
host_name: 2 values with extra whitespace
neighbourhood_group: 0 values with extra whitespace
neighbourhood: 0 values with extra whitespace
room_type: 0 values with extra whitespace


In [27]:
# Find possible text encoding artifacts

for col in ['name', 'host_name']:
    mask = df[col].astype(str).str.contains(r'Ã|Â|â', regex=True, na=False)
    print(f"\n{col}: {mask.sum()} possible encoding issues")

    if mask.sum() > 0:
        display(df.loc[mask, ['id', col]].head(10))


name: 891 possible encoding issues


,id,name
104,24143,"Williamsburg—Steps To Subway, Private Bath&B..."
132,30031,NYC artists’ loft with roof deck
202,54158,The Institute—Heart of Williamsburg
209,55467,Private Garden Apt • New Renovation
320,76761,Eveland the Place to Stay & Enjoy a 5-⭐️ 2...
351,94783,"Beautiful, Bright’s, Warm & Spacious 1.5BR Apt"
395,125053,âš¡Quiet Gem w/roof deck on NY's Hottest Stree...
398,126443,☆Massive DUPLEX☆ 2BR & 2BTH East Village 9...
464,163814,☆ STUDIO East Village ☆ Own bath! ☆ Slee...
481,168810,Lovely Vintage Haven—Heart of UWS



host_name: 236 possible encoding issues


,id,host_name
80,18764,Lulú
89,20611,Lulú
148,33223,Lulú
162,39704,Renée
258,63546,Antonín
262,63657,Renée
397,125594,Antonín
626,236788,Nick And Noémie
1558,705997,Katherine / François
1649,748656,Aimée


In [28]:
def fix_encoding(text):
    try:
        return text.encode('latin1').decode('utf-8')
    except:
        return text

df['name'] = df['name'].apply(fix_encoding)
df['host_name'] = df['host_name'].apply(fix_encoding)

print("Encoding cleanup completed.")

Encoding cleanup completed.


In [29]:
for col in ['name', 'host_name']:
    mask = df[col].astype(str).str.contains(r'Ã|Â|â', regex=True, na=False)
    print(f"{col}: {mask.sum()} issues remaining")

name: 652 issues remaining
host_name: 13 issues remaining


In [32]:
for col in ['name', 'host_name']:
    mask = df[col].astype(str).str.contains(r'Ã|Â|â', regex=True, na=False)
    print(f"\n{col}:")
    display(df.loc[mask, ['id', col]].head(10))


name:


,id,name
104,24143,"Williamsburg—Steps To Subway, Private Bath&B..."
132,30031,NYC artists’ loft with roof deck
202,54158,The Institute—Heart of Williamsburg
209,55467,Private Garden Apt • New Renovation
351,94783,"Beautiful, Bright’s, Warm & Spacious 1.5BR Apt"
395,125053,âš¡Quiet Gem w/roof deck on NY's Hottest Stree...
398,126443,☆Massive DUPLEX☆ 2BR & 2BTH East Village 9...
464,163814,☆ STUDIO East Village ☆ Own bath! ☆ Slee...
481,168810,Lovely Vintage Haven—Heart of UWS
579,219793,✿✿✿ COUNTRY COTTAGE IN THE CITY✿✿✿



host_name:


,id,host_name
11016,8507219,​ Valéria
11122,8612715,​ Valéria
21522,17224797,Élodie
27774,21786780,★The Local Apple★
35089,27827110,Öykü
41028,31898844,Edwin’S Place
41029,31900179,Edwin’S Place
44253,34126874,H Ai
45361,34702046,H Ai
45411,34723946,H Ai


In [33]:
# Detect outliers using the IQR method

numeric_cols = [
    'price',
    'minimum_nights',
    'number_of_reviews',
    'reviews_per_month',
    'calculated_host_listings_count',
    'availability_365'
]

for col in numeric_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outliers = ((df[col] < lower) | (df[col] > upper)).sum()

    print(f"{col}: {outliers} outliers")

price: 2972 outliers
minimum_nights: 6605 outliers
number_of_reviews: 6018 outliers
reviews_per_month: 3309 outliers
calculated_host_listings_count: 7073 outliers
availability_365: 0 outliers


## Outlier Detection and Treatment

Outliers were identified using the Interquartile Range (IQR) method.

Several numerical columns contained statistical outliers. However, these values were not automatically removed because unusual values may represent legitimate Airbnb listings or host behavior.

- `price`: High-priced listings may be legitimate.
- `minimum_nights`: Long-term stays may be valid.
- `number_of_reviews`: Popular listings can naturally have many reviews.
- `reviews_per_month`: High review activity can be legitimate.
- `calculated_host_listings_count`: Professional hosts may manage multiple listings.
- `availability_365`: No IQR outliers were detected.

Therefore, the statistical outliers were retained rather than capped or removed.

Listings with a price of 0 were handled separately and removed because a zero price is not meaningful for standard Airbnb price analysis.

In [34]:
# Final validation

print("Final dataset shape:", df.shape)

print("\nMissing values:")
display(df.isnull().sum()[df.isnull().sum() > 0])

print("\nDuplicate rows:", df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

Final dataset shape: (48884, 16)

Missing values:


,0
last_review,10051



Duplicate rows: 0

Data types:
id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                      int64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
dtype: object


In [35]:
# Save the cleaned dataset

df.to_csv('airbnb_cleaned.csv', index=False)

print("Cleaned dataset saved successfully as airbnb_cleaned.csv")

Cleaned dataset saved successfully as airbnb_cleaned.csv


In [36]:
from google.colab import files
files.download('airbnb_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Conclusion

In this project, I cleaned and prepared an Airbnb dataset by checking for missing values, duplicate records, incorrect data types, inconsistent text formatting, invalid values, and outliers.

I removed listings with zero prices, handled missing values based on the meaning of each column, converted the `last_review` column into the correct date format, and checked the dataset for unusual values using the IQR method.

The outliers were mostly retained because they could represent genuine Airbnb listings rather than errors. After cleaning, the dataset contains 48,884 rows with no duplicate records and is ready for further analysis.